File: home/home5.1/src/MegaDataPool.java
import java.util.ArrayList;
import java.util.List;

public class MegaDataPool {
    private final List<MegaData> pool;
    private final List<MegaData> available;

    public MegaDataPool(int size) {
        pool = new ArrayList<>(size);
        available = new ArrayList<>(size);
        for (int i = 0; i < size; i++) {
            MegaData data = new MegaData();
            pool.add(data);
            available.add(data);
        }
    }

    public synchronized MegaData acquire() {
        if (available.isEmpty()) {
            return null; 
        }
        return available.remove(available.size() - 1);
    }

    public synchronized void release(MegaData obj) {
        obj.reset();
        available.add(obj);
    }

    public int size() {
        return pool.size();
    }

    public int usedSize() {
        return pool.size() - available.size();
    }
}

File: class/practice5.1/MegaDataPool.cpp
#include "MegaDataPool.h"

MegaDataPool::MegaDataPool(size_t size) {
    pool.reserve(size);
    for (size_t i = 0; i < size; ++i) {
        pool.push_back(std::make_unique<MegaData>());
        available.push_back(pool.back().get());
    }
}

[[nodiscard]] MegaData* MegaDataPool::acquire() {
    if (available.empty()) {
        return nullptr; 
    }
    MegaData* obj = available.back();
    available.pop_back();
    return obj;
}

void MegaDataPool::release(MegaData* obj) {
    obj->reset();
    available.push_back(obj);
}

size_t MegaDataPool::size() const {
    return pool.size();
}

size_t MegaDataPool::usedSize() const {
    return pool.size() - available.size();
}

